pkg/aws/ec2.go

```diff
@@ -130,19 +130,25 @@ func (configMapInfo AwsDynamicConfig) GetInstanceAddress(kubeClient client.Clien
 	}
 	if len(res.Reservations) > 0 {
 		if len(res.Reservations[0].Instances) > 0 {
-			if res.Reservations[0].Instances[0].PublicDnsName != nil && *res.Reservations[0].Instances[0].PublicDnsName != "" {
+			instance := res.Reservations[0].Instances[0]
+			return configMapInfo.checkInstanceConnectivity(&instance, log)
+		}
+	}
+	return "", nil
+}
 
-				server, _ := net.ResolveTCPAddr("tcp", *res.Reservations[0].Instances[0].PublicDnsName+":22")
-				conn, err := net.DialTCP("tcp", nil, server)
-				if err != nil {
-					log.Error(err, "failed to connect to AWS instance")
-					return "", err
-				}
-				defer conn.Close()
+func (configMapInfo AwsDynamicConfig) checkInstanceConnectivity(instance *types.Instance, log *logr.Logger) (string, error) {
+	if instance.PublicDnsName != nil && *instance.PublicDnsName != "" {
 
-				return *res.Reservations[0].Instances[0].PublicDnsName, nil
-			}
+		server, _ := net.ResolveTCPAddr("tcp", *instance.PublicDnsName+":22")
+		conn, err := net.DialTCP("tcp", nil, server)
+		if err != nil {
+			log.Error(err, "failed to connect to AWS instance")
+			return "", err
 		}
+		defer conn.Close()
```

Comment:
Is there no connection to close if the error is non-nil?

Reply:
no, if you get an error you don't get a connection